server/router/notification.ts

```diff
@@ -58,99 +57,108 @@ const sendMsg = (sub: Subscription, msg: string) => {
+  const timeAdded = (new Date()).toISOString();
+  if (platform === PlatformType.WEB) {
     const subscription = {
-      platform: plat,
+      id: req.webSub!.endpoint + userType + platform,
       endpoint: req.webSub!.endpoint,
-      keys: req.webSub!.keys,
+      userType,
+      platform,
+      timeAdded,
+      preferences: [],
+      keys: req.webSub!.keys, // TODO user id to user
+    };
+    addSub(subscription).then(() => resolve('success')).catch(reject);
+  } else {
+    const snsParams = {
+      Token: req.token!,
+      PlatformApplicationArn: snsValues.android,
     };
-    subscriptionSet.add(JSON.stringify(subscription));
-    return new Promise((resolve, reject) => {
-      resolve('success');
-    });
-  }
-
-  const ios = plat === 'ios rider' ? snsValues.ios_rider : snsValues.ios_driver;
-  const arn = plat === 'android' ? snsValues.android : ios;
-  const snsParams = {
-    Token: req.token!,
-    PlatformApplicationArn: arn,
-  };
-
-  return new Promise((resolve, reject) => {
     sns.createPlatformEndpoint(snsParams, (err, data) => {
       if (err || !data) {
         reject();
-        return;
+      } else {
+        const subscription = {
+          id: data.EndpointArn + userType + platform,
+          endpoint: data.EndpointArn!,
+          userType,
+          platform,
+          timeAdded,
+          preferences: [],
+        };
+        addSub(subscription).then(() => resolve('success')).catch(reject);
       }
-
-      const subscription = {
-        platform: plat,
-        endpoint: data.EndpointArn,
-      };
-      subscriptionSet.add(JSON.stringify(subscription));
-      resolve('success');
     });
-  });
-};
+  }
+});
 
 // send out a notification to everyone
 router.post('/sendAll', (req, res) => {
   const { msg } = req.body; // TODO validation, error codes
-
-  const promises = [...subscriptionSet].map((strSub) => {
-    const sub = JSON.parse(strSub);
-    return sendMsg(sub, msg);
+  Subscription.scan().exec((err, data) => {
+    if (err || !data) {
+      res.status(400).send({ err: 'could not get all subscriptions' });
+    } else {
+      const promises = data.map((doc) => {
+        const sub = JSON.parse(JSON.stringify(doc.toJSON()));
+        return sendMsg(sub, msg);
+      });
+      Promise.all(promises)
+        .then(() => {
+          res.status(200).send({ success: true });
+        })
+        .catch(() => {
+          res.status(500).send({ err: 'could not send all messages' });
+        });
+    }
   });
-  Promise.all(promises)
-    .then((data) => {
-      res.status(200).send({ success: true });
-    })
-    .catch((err) => {
-      res.status(500).send({ err: 'could not send all messages' });
-    });
 });
 
 // subscribe a user
 router.post('/subscribe', validateUser('User'), (req, res) => {
-  if (badPlatform(req.body.platform)) {
+  const { platform } = req.body;
+  if (platform !== PlatformType.WEB && platform !== PlatformType.ANDROID) {
```

Comment:
If I try to subscribe on an iOS device, wouldn't this result in an invalid platform error?

Reply:
We are using GCM for both android and IOS right now, so the platform type is ANDROID for both, which is a bit confusing.